service/controller/resource/nodepool/create.go

```diff
@@ -0,0 +1,94 @@
+
+	"github.com/giantswarm/azure-operator/v4/pkg/annotation"
+	"github.com/giantswarm/azure-operator/v4/service/controller/internal/state"
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+	"github.com/giantswarm/azure-operator/v4/service/controller/resource/masters"
+)
+
+// createStateMachine configures and returns state machine that is driven by
+// EnsureCreated.
+func (r *Resource) createStateMachine() state.Machine {
+	sm := state.Machine{
+		Logger:       r.Logger,
+		ResourceName: Name,
+		Transitions: state.TransitionMap{
+			DeploymentUninitialized:     r.deploymentUninitializedTransition,
+			ScaleUpWorkerVMSS:           r.scaleUpWorkerVMSSTransition,
+			CordonOldWorkers:            r.cordonOldWorkersTransition,
+			WaitForWorkersToBecomeReady: r.waitForWorkersToBecomeReadyTransition,
+			DrainOldWorkerNodes:         r.drainOldWorkerNodesTransition,
+			TerminateOldWorkerInstances: r.terminateOldWorkersTransition,
+			ScaleDownWorkerVMSS:         r.scaleDownWorkerVMSSTransition,
+		},
+	}
+
+	return sm
+}
+
+// This resource applies the ARM template for the worker instances, monitors the process and handles upgrades.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	azureMachinePool, err := key.ToAzureMachinePool(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if isMasterUpgrading(&azureMachinePool) {
+		r.Logger.LogCtx(ctx, "level", "debug", "message", "master is upgrading")
+		r.Logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+		return nil
+	}
+
+	var newState state.State
+	var currentState state.State
+	{
+		s, err := r.getCurrentState(ctx, azureMachinePool)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		currentState = state.State(s)
+
+		r.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("current state: %s", currentState))
+		newState, err = r.StateMachine.Execute(ctx, obj, currentState)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	err = r.CtrlClient.Status().Update(ctx, &azureMachinePool)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if newState != currentState {
+		r.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("new state: %s", newState))
+		r.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("setting resource status to %#q", newState))
+		err = r.saveCurrentState(ctx, azureMachinePool, string(newState))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		r.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("set resource status to %#q", newState))
+		r.Logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation")
+		reconciliationcanceledcontext.SetCanceled(ctx)
+	} else {
+		r.Logger.LogCtx(ctx, "level", "debug", "message", "no state change")
+	}
+
+	return nil
+}
+
+func isMasterUpgrading(getter key.AnnotationsGetter) bool {
+	masterUpgrading, exists := getter.GetAnnotations()[annotation.IsMasterUpgrading]
```

Comment:
I think we need to revisit this in the near future. Having annotation of master[s] state in every `AzureMachinePool` CR is not ideal. I'd expect to find the state of masters in `AzureMachine[s]` of master nodes or alternatively in `AzureCluster` (maybe better since that would be single common location).

Maybe create a separate issue to revisit this, add a comment here and continue with this one?